File: apps/user/lib/models/security_incident.dart
class SecurityIncident {
  SecurityIncident({
    required this.id,
    required this.title,
    required this.severity,
    required this.status,
    required this.detectedAt,
    required this.timeline,
    this.detectionSource,
    this.impactSummary,
    this.mitigationSteps,
  });

  factory SecurityIncident.fromJson(Map<String, dynamic> json) {
    final timeline = (json['timeline'] as List<dynamic>? ?? [])
        .whereType<Map>()
        .map((entry) => SecurityIncidentTimelineEntry.fromJson(
            Map<String, dynamic>.from(entry)))
        .toList();

    return SecurityIncident(
      id: json['id'] as String,
      title: json['title'] as String? ?? '',
      severity: json['severity'] as String? ?? 'medium',
      status: json['status'] as String? ?? 'open',
      detectionSource: json['detection_source'] as String?,
      impactSummary: json['impact_summary'] as String?,
      mitigationSteps: json['mitigation_steps'] as String?,
      detectedAt: DateTime.tryParse(json['detected_at'] as String? ?? '') ?? DateTime.now(),
      timeline: timeline,
    );
  }

  final String id;
  final String title;
  final String severity;
  final String status;
  final DateTime detectedAt;
  final String? detectionSource;
  final String? impactSummary;
  final String? mitigationSteps;
  final List<SecurityIncidentTimelineEntry> timeline;

  String get severityLabel => severity.toUpperCase();

  String get statusLabel =>
      status.isEmpty ? '' : status[0].toUpperCase() + status.substring(1);
}

class SecurityIncidentTimelineEntry {
  SecurityIncidentTimelineEntry({
    required this.event,
    required this.occurredAt,
    this.actorId,
    this.note,
  });

  factory SecurityIncidentTimelineEntry.fromJson(Map<String, dynamic> json) {
    final context = json['context'] is Map
        ? Map<String, dynamic>.from(json['context'] as Map)
        : const <String, dynamic>{};

    return SecurityIncidentTimelineEntry(
      event: json['event'] as String? ?? '',
      occurredAt: DateTime.tryParse(json['occurred_at'] as String? ?? '') ?? DateTime.now(),
      actorId: context['actor_id'] as int?,
      note: context['note'] as String?,
    );
  }

  final String event;
  final DateTime occurredAt;
  final int? actorId;
  final String? note;
}

File: app/Models/SecurityIncident.php
<?php

namespace App\Models;

use App\Enums\SecurityIncidentSeverity;
use App\Enums\SecurityIncidentStatus;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Support\Arr;
use Illuminate\Support\Str;

class SecurityIncident extends Model
{
    use HasFactory;

    protected $fillable = [
        'public_id',
        'title',
        'severity',
        'status',
        'detection_source',
        'impacted_assets',
        'timeline',
        'impact_summary',
        'mitigation_steps',
        'root_cause',
        'follow_up_actions',
        'runbook_updates',
        'detected_at',
        'acknowledged_at',
        'resolved_at',
        'created_by',
        'acknowledged_by',
        'resolved_by',
    ];

    protected $casts = [
        'impacted_assets' => 'array',
        'timeline' => 'array',
        'follow_up_actions' => 'array',
        'runbook_updates' => 'array',
        'detected_at' => 'datetime',
        'acknowledged_at' => 'datetime',
        'resolved_at' => 'datetime',
    ];

    protected static function booted(): void
    {
        static::creating(function (self $incident) {
            if (blank($incident->public_id)) {
                $incident->public_id = (string) Str::uuid();
            }

            if (! in_array($incident->severity, SecurityIncidentSeverity::values(), true)) {
                $incident->severity = SecurityIncidentSeverity::MEDIUM->value;
            }

            if (! in_array($incident->status, SecurityIncidentStatus::values(), true)) {
                $incident->status = SecurityIncidentStatus::OPEN->value;
            }
        });
    }

    public function reporter(): BelongsTo
    {
        return $this->belongsTo(User::class, 'created_by');
    }

    public function acknowledger(): BelongsTo
    {
        return $this->belongsTo(User::class, 'acknowledged_by');
    }

    public function resolver(): BelongsTo
    {
        return $this->belongsTo(User::class, 'resolved_by');
    }

    public function appendTimeline(string $event, array $context = []): void
    {
        $timeline = Arr::wrap($this->timeline);

        $timeline[] = [
            'event' => $event,
            'context' => $context,
            'occurred_at' => now()->toIso8601String(),
        ];

        $this->timeline = $timeline;
    }

    public function getRouteKeyName(): string
    {
        return 'public_id';
    }

    public function severity(): SecurityIncidentSeverity
    {
        return SecurityIncidentSeverity::from($this->severity);
    }

    public function status(): SecurityIncidentStatus
    {
        return SecurityIncidentStatus::from($this->status);
    }
}
